ENH: set empty array norm to zero.

ogauthe opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
closes #12534

#### What does this implement/fix?
For `ord=None`, scipy.linalg.norm crashes when a 0-size array is given as input with a very confusing message:
```
In [1]: scipy.linalg.norm([])                                                                                                                     
---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
<ipython-input-1-07a65e21029c> in <module>
----> 1 scipy.linalg.norm([])

~/Library/miniconda3/lib/python3.7/site-packages/scipy/linalg/misc.py in norm(a, ord, axis, keepdims, check_finite)
    148             # use blas for fast and stable euclidean norm
    149             nrm2 = get_blas_funcs('nrm2', dtype=a.dtype)
--> 150             return nrm2(a)
    151 
    152         if a.ndim == 2 and axis is None and not keepdims:

error: (offx>=0 && offx<len(x)) failed for 2nd keyword offx: dnrm2:offx=0
```

The easy solution is to set the norm of an empty array to 0. This is already the case for `numpy.linalg.norm` for `ord=fro`. However, for `ord<=0`, the result is not a mathematical norm and is hard to define for empty arrays. Here I propose to raise an error in such cases.
#### Additional information
<!--Any additional information you think is important.-->

tylerjereddy commented on scipy/linalg/misc.py:
```diff
@@ -141,6 +142,13 @@ def norm(a, ord=None, axis=None, keepdims=False, check_finite=True):
+    # define empty array norm as 0.0 for ord > 0
+    if a.size == 0:
+        if ord in (None, "fro") or ord > 0:
+            return 0.0
```
Seems to be consistent with `np.linalg.norm([])`, as you say. Perhaps once a linalg expert weighs in the behavior might be encoded in a regression test as well.

rgommers replied:
Argh, `scipy.linalg.norm` doesn't seem to have any tests as far as I can tell, it's only used inside tests for other functionality.

@ogauthe thanks for working on this. Could you add a `scipy/linalg/tests/test_misc.py` file, and add one regression test for this size-0 behavior to it?

WarrenWeckesser replied:
Returning the scalar 0 is not correct if the shape of `a` is nontrivial and an `axis` has been given.  For example, if `z` has shape `(3, 0, 2)`, and `axis=1` is given, the result should be an array with shape `(3, 2)` if `keepdims` is False.  If `keepdims` is True, the shape should be `(3, 1, 2)`.  E.g.

```
In [95]: z = np.empty((3, 0, 2))

In [96]: z
Out[96]: array([], shape=(3, 0, 2), dtype=float64)

In [97]: z.size
Out[97]: 0

In [98]: np.linalg.norm(z, axis=1)
Out[98]: 
array([[0., 0.],
       [0., 0.],
       [0., 0.]])

In [99]: np.linalg.norm(z, axis=1, keepdims=True)
Out[99]: 
array([[[0., 0.]],

       [[0., 0.]],

       [[0., 0.]]])
```

ilayn commented:
I didn't even know that we had a separate implementation. I thought this is just an alias from numpy. But apparently not.

ogauthe commented:
@ilayn maybe having it as an alias would be a better idea. The more I think about it the more I'm convinced. There is little reason to maintain two versions of a function doing the same thing, and in most cases scipy falls back to numpy.linalg.norm. More, it means the documentation is actually outdated. The scipy function supports `ord=nuc` because it call `numpy.linalg.norm` which supports it, but this is not documented by scipy.

Currently there are two main differences:
- the use of blas nrm2 and lange in some cases (numpy should give the same results)
- the keyword `check_finite`, which does not exist in numpy. This is more a problem since it changes the function signature.
I think this PR is good to take (I can submit the same in numpy), but one should wonder if the differences are worth it. If not, `check_finite` should be deprecated before making this function an alias. 

@rgommers I will try to write a test for this.

WarrenWeckesser commented:
FYI: Currently, we can't make `scipy.linalg.norm` an alias for `numpy.linalg.norm`.  In `scipy.stats.pearsonr` (specifically [here](https://github.com/scipy/scipy/blob/64a342f9c20453d4b5d5363d9aba435a53e92db7/scipy/stats/stats.py#L4034)), I intentionally used `scipy.linalg.norm` instead of `numpy.linalg.norm` because the numpy version suffers from overflow when the arguments are large:
```
In [26]: import scipy.linalg

In [27]: x = np.array([3e160, 4e160])

In [28]: scipy.linalg.norm(x)
Out[28]: 5e+160

In [29]: np.linalg.norm(x)
Out[29]: inf
```
I think this is a defect in `numpy.linalg.norm`, but I don't know if there is interest in fixing it--it might introduce extra overhead that hurts performance.

ogauthe commented:
meh. So using `nrm2` makes a real difference in some cases and I forgot to consider `axis`. Still some work to do so.

WarrenWeckesser commented:
I suspect the easiest way to handle this is to consider `a.size == 0` as another case where we let `numpy.linalg.norm` do the work.  That is, the change might be as simple including `a.size != 0` in the conditions required for using an "optimized" norm:

    # Only use optimized norms if axis and keepdims are not specified.
    if a.dtype.char in 'fdFD' and axis is None and not keepdims and a.size != 0:

ogauthe commented:
So I changed several things.
- the 4 last versions of numpy support `axis` and `keepdim`, so I removed old code needed if they were not supported
- they also support `ord=nuc`, which I added to the documentation
- I followed @WarrenWeckesser 's advice and I let numpy deal with 0 size arrays
- I cleaned the documentation and the code

ogauthe commented:
added `scipy/linalg/tests/test_misc.py`. Not sure if something more is needed as a unitary test.

mckib2 commented:
@WarrenWeckesser @ilayn This PR looks pretty close, I'll kick off the CI -- have all the comments been resolved?  A squash commit seems reasonable for this changeset

WarrenWeckesser commented on scipy/linalg/misc.py:
```diff
@@ -142,14 +147,14 @@ def norm(a, ord=None, axis=None, keepdims=False, check_finite=True):
         a = np.asarray(a)
 
     # Only use optimized norms if axis and keepdims are not specified.
```
This comment is now even more out of date--it doesn't mention the dtype or the size.  Let's just remove it--it doesn't add any information that isn't clear from the condition in the `if`-statement.
```suggestion
```

WarrenWeckesser commented on scipy/linalg/misc.py:
```diff
@@ -62,23 +65,25 @@ def norm(a, ord=None, axis=None, keepdims=False, check_finite=True):
+    The nuclear norm is the sum of the singular values.
+
+    Both the Frobenius and nuclear norm orders are only defined for
+    matrices and raise a ValueError when `a.ndim != 2`.
```
For these norms, `a.ndim > 2` is allowed when used in combination with `axis`, so this statement isn't strictly true.  We could just drop the second part of the sentence:
```suggestion
    matrices.
```

WarrenWeckesser reviewed (changes requested):
Looks good!  I have two small suggested changes.

ogauthe commented:
Your points are relevant, I made the suggested changes.

WarrenWeckesser commented on scipy/linalg/tests/test_misc.py:
```diff
@@ -0,0 +1,4 @@
+from scipy.linalg import norm
+
+def test_norm():
```
Just noticed one more tiny issue: PEP-8 requires two blank line before a module-level `def`:
```suggestion

def test_norm():
```

WarrenWeckesser reviewed (approved):
Thanks @ogauthe!  I found one more tiny issue, but I'll fix that myself and merge.

WarrenWeckesser commented:
Merged, thanks @ogauthe.
